src/main/java/roboy/dialog/DialogStateMachine.java

```diff
@@ -0,0 +1,545 @@
+        activeState = null;
+        rosMainNode = rmn;
+        memory = mem;
+        optionalPersFileInfo = new HashMap<>();
+
+        if (rosMainNode == null) {
+            logger.info("RosMainNode will be unavailable in DialogStateMachine (null was passed)");
+        }
+        if (memory == null) {
+            logger.info("Memory will be unavailable in DialogStateMachine (null was passed)");
+        }
+    }
+
+    /**
+     * Create an empty OFFLINE DialogStateMachine without a reference to the RosMainNode and Memory.
+     * States will not be able to access the RosMainNode and Memory functionality.
+     * This constructor is mainly used for testing.
+     */
+    public DialogStateMachine() {
+        this(null, null);
+    }
+
+    //endregion
+
+
+    // #####################################################
+    // #                 Getter & Setter                   #
+    // #####################################################
+
+    //region initial state
+
+    /**
+     * Returns the initial state for this state machine.
+     * @return initial state for this state machine
+     */
+    public State getInitialState() {
+        return initialState;
+    }
+
+    /**
+     * Set the initial state of this state machine.
+     * The state will be automatically added to the machine if not already added.
+     * If active state was null, it will be set to the new initial state.
+     * @param initial initial state
+     */
+    public void setInitialState(State initial) {
+        if (!identifierToState.containsValue(initial) && initial != null) {
+            addState(initial);
+        }
+        this.initialState = initial;
+        if (activeState == null) { // if active state is not set yet, aet active to initial
+            setActiveState(initial);
+        }
+    }
+
+    /**
+     * Set the initial state of this state machine using state identifier.
+     * If there is no state with specified identifier, you will get an error message
+     * and the initial state will be set to null.
+     * @param identifier identifier of the state that should become the initial state
+     */
+    public void setInitialState(String identifier) {
+        State initial = identifierToState.get(identifier);
+        if (initial == null) {
+            logger.error("setInitialState(" + identifier + "): Unknown identifier!");
+        }
+        setInitialState(initial);
+    }
+
+    //endregion
+
+    //region active state
+
+    /**
+     * Returns the active state for this state machine.
+     * @return active state for this state machine
+     */
+    public State getActiveState() {
+        return activeState;
+    }
+
+    /**
+     *
+     * Set the active state of this state machine.
+     * The state will be automatically added to the machine if not already added.
+     * @param s state to make active
+     */
+    public void setActiveState(State s) {
+        if (!identifierToState.containsValue(s) && s != null) {
```

Comment:
Maybe the other way around as there is no need to call the method if s == null

Reply:
good point!